### ISSUE TITLE
componentWillReceiveProps() receives undefined instead of default

### ISSUE THREAD

**Issue by TrevorBurnham** [airbnb/enzyme#1628]
### Current behavior

Calling `wrapper.setProps({prop: undefined})` (where `wrapper` is a shallow wrapper) causes the component's `componentWillReceiveProps()` to be called with `{prop: undefined}`, even if `prop` has a default value.

### Expected behavior

The component's `componentWillReceiveProps()` should be called with `{prop: propDefault}` (where `propDefault` is the default value of the prop), to reflect the "real-world" behavior.

#### API

- [x] shallow
- [ ] mount
- [ ] render

#### Version

| library         | version
| ---------------- | -------
| Enzyme           | 3.3.0
| React            |  15.3.1

#### Adapter

- [ ] enzyme-adapter-react-16
- [ ] enzyme-adapter-react-15
- [x] enzyme-adapter-react-15.4
- [ ] enzyme-adapter-react-14
- [ ] enzyme-adapter-react-13
- [ ] enzyme-adapter-react-helper
- [ ] others ( ) 

---

I've included a test case below. Note that all tests pass if you use `mount()` instead of `shallow()`, demonstrating the "real-world" lifecycle event behavior.

```jsx
import React from 'react';
import { shallow } from 'enzyme';

const Component = React.createClass({
  getDefaultProps() {
    return { className: 'default-class' };
  },

  componentWillReceiveProps(nextProps) {
    this._receivedClassName = nextProps.className;
  },

  render() {
    return <div className={this.props.className} />;
  },
});

describe('Component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Component />);
  });

  it('renders with the default className', () => { // PASS
    expect(wrapper.prop('className')).toBe('default-class');
  });

  describe('when className is explicitly set to undefined', () => {
    beforeEach(() => {
      wrapper.setProps({ className: undefined });
    });

    it('renders with the default className', () => { // PASS
      expect(wrapper.prop('className')).toBe('default-class');
    });

    it('gets the default className in componentWillReceiveProps()', () => { // FAIL
      expect(wrapper.instance()._receivedClassName).toBe('default-class');
    });
  });
});
```

**Comment by mykhailo-riabokon** [comment]
Hi @TrevorBurnham. It seems this behaviour can be reproduced with latest React and React adapter as well. [setProps method](https://github.com/airbnb/enzyme/blob/master/packages/enzyme/src/ShallowWrapper.js#L255) for `shallow` wrapper combines prevProps (which includes defaultProps) and props which you pass to the `setProps` function, and if you have case like:

```javascript
const prevProps = { className: 'default-class' };
const props = { className: undefined };
const nextProps = { ...prevProps, ...props } // { className: undefined } as className was overwritten
```
When you use `mount` wrapper basically React.createElement is used, and it has [this check](https://github.com/facebook/react/blob/master/packages/react/src/ReactElement.js#L222) which sets prop value. 

@ljharb what do you think, is it a bug? If so could try to create a PR with possible fix.

**Comment by ljharb** [comment]
@mykhailo-riabokon this does seem like a bug, and a PR would be very helpful!

**Comment by ljharb** [comment]
This seems fixed in master; #1711 has passing tests.
